### ISSUE TITLE
theme.*Resource wrapper types call svg.Colorize() on raster resources

### ISSUE THREAD

**Issue by davidhaymond** [fyne-io/fyne#3896]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

When using raster images as resources such as button icons, they are assumed to be SVG, causing errors like the following to be logged:

```
2023/05/19 11:10:44 Fyne error:  could not load SVG, falling back to static content:
2023/05/19 11:10:44   Cause: XML syntax error on line 3: invalid UTF-8
2023/05/19 11:10:44   At: /home/david/go/pkg/mod/fyne.io/fyne/v2@v2.3.4/internal/svg/svg.go:27
```

This happens because the `Content()` methods of the `theme.*Resource` wrapper types in `theme/icons.go` pass the content of the underlying resource to `svg.Colorize()` without checking to see if the underlying resource is SVG. This causes `svg.Colorize()` to log XML parsing errors. The following types have this problem:

- `theme.ThemedResource'
- `theme.InvertedThemedResource'
- `theme.ErrorThemedResource'
- `theme.PrimaryThemedResource'
- `theme.DisabledResource'

### How to reproduce

1. Create a button with a raster icon.
2. Cause one of the themed resource wrapper types to be used (e.g. disable the button to cause the icon to be wrapped in a `theme.DisabledResource`).

See the example code below.

### Example code

```go
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	w := a.NewWindow("Hello World")

	icon, _ := fyne.LoadResourceFromURLString("https://mirrors.creativecommons.org/presskit/icons/cc.png")
	btn := widget.NewButtonWithIcon("", icon, func() {})
	btn.Disable()
	w.SetContent(btn)
	w.ShowAndRun()
}
```

### Fyne version

2.3.4

### Go compiler version

1.20.4

### Operating system and version

EndeavorOS

**Comment by andydotxyz** [comment]
This is an interesting point.
I wonder if perhaps logging an error is correct, but that it should be more appropriate.
Otherwise won't we get more bugs opened to say the raster icon is not disabling when a button is disabled? :)

**Comment by davidhaymond** [comment]
I agree that a more appropriate error would be helpful, and maybe also a note in the documentation that only SVG icons are supported by the theming system. However, SVG may not be an option in some cases, and that causes a lot of log noise. Is there a case to be made for supporting themed raster resources? If that isn't the direction you wish to go in, I can at least draft a PR that improves the error messages.

**Comment by dweymouth** [comment]
I would argue that logging an error is incorrect, in the case that the developer is using a raster icon for a button. Perhaps the API documentation should make more clear that while PNGs are supported for icons, they will not update their appearance with theming changes. Or (and this is a bigger task), Fyne apps should have a log level setting (either at build time or with an env variable), where by default logs like these would be suppressed unless debug logging is turned on.

This originally came up as an issue report in my Supersonic project, where users interpreted the error logs as indicative of a bug with the app (or in any case, that they are just polluting and non-useful). While it would be nice to use SVG icons for everything, small projects that don't have designers on board might have to fall back to PNGs occasionally, and having these messages log out every time a button is enabled/disabled or the theme setting is changed is a little polluting IMO.

**Comment by andydotxyz** [comment]
> It's not an error condition, it's the expected behavior with raster icons.

I guess that is the crux of the matter - do developers expect that it won't be changed if not an SVG?

I know some other toolkits will re-colour bitmap resources in similar situations.

I suppose the question is indeed what is supported and what expectation should we set - that will determine what is an error and what is not.

**Comment by andydotxyz** [comment]
> Fyne apps should have a log level setting (either at build time or with an env variable), where by default logs like these would be suppressed unless debug logging is turned on.

We already have "debug" and "hint" modes to do exactly that.

**Comment by andydotxyz** [comment]
> While it would be nice to use SVG icons for everything, small projects that don't have designers on board might have to fall back to PNGs occasionally

Honestly I never correlated SVG with having a designer. It was never true during the creation of Fyne. There is a massive repository of Material Design icons that can be downloaded in SVG format and that has worked for most apps.

**Comment by davidhaymond** [comment]
> I suppose the question is indeed what is supported and what expectation should we set - that will determine what is an error and what is not.

If developers wish to extend the resource system to support themable raster resources, as @dweymouth has done, there is no way to opt out of the error messages. Personally, I think the documentation should state that only SVG resources can be themed, but still allow and encourage developers to add custom raster support on top without penalizing them with error messages.

Alternatively, perhaps there can be a way to differentiate between themable and unthemable resources in the type system? Only the themable resource types would log an error for raster contents. That would allow the developer to make explicit their intention to "opt out" of the theming system (e.g. an `UnthemableResource`, which makes clear that it will never change color without manual intervention).

**Comment by junmaqiang** [comment]
meby this issues same
https://github.com/fyne-io/fyne/issues/3900

**Comment by andydotxyz** [comment]
No, one refers to SVGs not working and another refers to rasters being assumed to be SVG.
